shell/server/proxy.js

```diff
@@ -36,6 +36,85 @@ SANDSTORM_ALTHOME = Meteor.settings && Meteor.settings.home;
+  // update Bob's Identity to say it's referredBy Alice's Account (which is apiTokenAccountId).
+  check(accountId, String);
+  check(identityId, String);
+  check(apiTokenAccountId, String);
+
+  // Bail out early if quota enforcement is disabled.
+  if (! Meteor.settings.public.quotaEnabled) {
+    return;
+  }
+
+  var aliceAccountId = apiTokenAccountId;
+  var bobAccountId = accountId;
+  var bobIdentityId = identityId;
+
+  if (Meteor.users.find({_id: bobAccountId,
+                         referredByComplete: {$exists: true}}).count() > 0) {
+    return;
+  }
+
+  // Only actually update Bob's Identity ID if there is no referredBy.
+  var updatedCount = Meteor.users.update(
+    {_id: bobIdentityId, referredBy: {$exists: false}},
+    {$set: {referredBy: aliceAccountId}});
+}
+
+function referralProgramLogSharingTokenUse(db, bobAccountId) {
+  // Hooray! The sharing token is valid! Someone (let's call them Charlie) is going to get a UiView
+  // to this grain!  This means that the user who created this apiToken knows how to use the "share
+  // access" interface. Let's call them Bob.
+  //
+  // If Bob's Account.referredByComplete is not yet set, then look at Bob's Identities and take the
+  // first referredBy we find -- let's call that Alice.
+  //
+  // Copy Alice's account ID Account.referredByComplete, and then update Alice's referredIdentityIds
+  // to point at Bob's Identity, and then remove the referredBy from Bob's Identity since it has
+  // become redundant.
+  //
+  // Implementation note: this does mean that Alice can get referral credit for Bob by sharing a
+  // link with Bob, even if Bob already had an account.
+
+  // Bail out early if quota support is not enabled.
+  if (! Meteor.settings.public.quotaEnabled) {
+    return;
+  }
+
+  // Bail out if Bob has a referredByComplete.
+  if (Meteor.users.find({_id: bobAccountId, referredByComplete: {$exists: true}}).count() > 0) {
+    return;
+  }
+
+  // Look for a referredBy on any of Bob's identities.
+  var identities = SandstormDb.getUserIdentities(Meteor.users.findOne({_id: bobAccountId}));
+  var identitiesWithReferredBy = _.filter(identities, function(identity) {
+    return !! identity.referredBy;
+  });
+  if (identitiesWithReferredBy.length === 0) {
+    return;
+  }
+
+  var bobIdentity = identitiesWithReferredBy[0];
+  var aliceAccountId = bobIdentity.referredBy;
+
+  // Store Bob's Account.referralCompletedBy.
+  var now = new Date();
+  Meteor.users.update({_id: bobAccountId, referredByComplete: {$exists: false}},
+                      {$set: {referredByComplete: bobIdentity.referredBy,
+                              referredCompleteDate: now}});
+
+  // Update Alice's Account.referredIdentityIds.
+  Meteor.users.update({_id: aliceAccountId},
+                      {$push: {referredIdentityIds: bobIdentity._id}});
+
+  // Remove now-redundant Bob identity referredBy.
```

Comment:
If I understand correctly, there's no particular need to remove this, so as mentioned above I'd leave it in and avoid unnecessary rewrites of the userdoc, plus you can save a query above.